### ISSUE TITLE
VRTK_SDKManager.persistOnLoad considered harmful

### ISSUE THREAD

**Issue by bddckr** [thestonefox/VRTK#1316]
# What?!

`VRTK_SDKManager.persistOnLoad` is a nice shortcut to `DontDestroyOnLoad` the game object the SDK Manager is on.

When Unity introduced [Multi Scene Editing](https://docs.unity3d.com/Manual/MultiSceneEditing.html) usage of [`Object.DontDestroyOnLoad`](https://docs.unity3d.com/ScriptReference/Object.DontDestroyOnLoad.html) has kinda been deprecated, as we can see in the first link:

> It is recommended to avoid using DontDestroyOnLoad to persist manager GameObjects that you want to survive across scene loads. Instead, create a manager scene that has all your managers and use [`SceneManager.LoadScene(<path>, LoadSceneMode.Additive)`](https://docs.unity3d.com/ScriptReference/SceneManagement.SceneManager.LoadScene.html) and [`SceneManager.UnloadScene`](https://docs.unity3d.com/ScriptReference/SceneManagement.SceneManager.UnloadScene.html) to manage your game progress.

With `VRTK_SDKManager.persistOnLoad` we now hit one of the reasons for that recommendation:
When the `DontDestroyOnLoad` call is done Unity moves the game object (and its children) to a new scene called `DontDestroyOnLoad`:

> In Playmode, whith [_sic_] multiple scenes in the Hierarchy, an additional scene will show up called DontDestroyOnLoad.
> 
> Prior to Unity 5.3, any objects you would instantiate in Playmode marked as “DontDestroyOnLoad” would still show up in the hierarchy. These objects are not considered part of any scene but for Unity to still show the objects, and for you to inspect them, these objects are now shown as part of the special DontDestroyOnLoad scene.
> 
> You do not have access to the DontDestroyOnLoad scene and it is not available at runtime.

The last sentence is important here, because we found out even a simple call like 
```C#
gameObjectThatIsInDontDestroyOnLoadScene.GetComponent<ComponentThatIsOnThatGameObject>()
```
results in `null`, despite that component actually being on the game object!

# So how do we fix this?

We follow Unity's suggestion, of course! That is: Using additive scenes instead.

**Because we know additive scenes are probably something that is still new to tons of our users we want to emphasize: The easiest solution for now is to just use the same SDK Manager & SDK Setups in _every_ scene. This means duplicated game objects of course, but it's easier than using additive scenes.**

Let's see how we can use additive scenes then:
- Move SDK Manager with SDK Setups into a new (empty) scene.
- Load that scene additively (using `LoadSceneMode.Additive`).
- Load any other scene additively, too.
- If there is a need to unload a scene do so via `SceneManager.UnloadScene`.

**Comment by bddckr** [comment]
PR to fix #1317.

**Comment by thestonefox** [comment]
Could the issue being caused by

https://github.com/thestonefox/VRTK/commit/1b179291038b41443dbfaeaa3ec416edc8a0f1d1

thinking is it's trying to load a previous SDK set up but there is no previous one, therefore if we bypassed the try previous reload if persist on load is ticked, something like:

```
public void TryLoadSDKSetupFromList(bool tryUseLastLoadedSetup = true)
{
  if (persistOnLoad)
  {
    tryUseLastLoadedSetup = false;
  }
..
..
..
}
```

**Comment by thestonefox** [comment]
I think actually the problem is with auto populate.

If you remove auto populate and have the objects manually set, then use persist on load it keeps the objects when switching scene.

**Comment by thestonefox** [comment]
Steps to recreate:

 * Open any example scene (e.g. 005)
 * On the SDK Manager tick `persist on load`
 * On the SDK Setup leave `auto populate` ticked
 * Run the scene and all of the SDK linked objects disappear
 * Return to the editor
 * Untick `auto populate` on the SDK Setup and run the scene
 * The SDK Linked objects stay
 * You can now switch to another scene that doesn't have an SDK setup and it works

**Comment by thestonefox** [comment]
Git Bisect determines this is the commit that broke `persist on load`

https://github.com/thestonefox/VRTK/commit/46c39396a21591bda2425c2cb15f38be0ff8b93a

**Comment by thestonefox** [comment]
The issue is with the fix above it ensures the objects are repopulated again so scripts can be moved to the correct parent (e.g. the controller alias objects).

However, if you're using persist on load then it nulls out the objects again but they're not in the current scene to find (as mentioned in the original post by @bddckr )

If you simply bypass the `PopulateObjectReferences` call in `OnEnable` if `persist on load = true` then the references stay but the auto parenting no longer happens again.

**Comment by thestonefox** [comment]
@trumpetx @bddckr @reznovVR 

My conclusions in this research is that the original suggestion of `persist on load` being unsuitable is valid and we should move away from using it.

**Comment by trumpetx** [comment]
At this point, I've taken the day away from making game stuff and changed to no longer use persist on load.  This is the second time the feature has bit me and quite frankly, I'm tired of fixing my project (instead of adding to it).  I don't have a dog in the fight, if persist on load is bad, I agree we should move away (I still don't quite understand why it's so breaking as it's a very common Unity thing - but I haven't really taken the time to learn so I accept your (and @bddckr 's) researched opinion).  

My only request is that it no longer be an option for VRTK.  There's no reason to have a feature that is fundamentally broken and breaks all expectation of what's happening.  As such, I support @bddckr 's direction in this PR to remove the checkbox and deprecate the feature post haste.

**Comment by bddckr** [comment]
@thestonefox What you figured out is pretty much what I saw: Even with the auto populate disabled it won't work completely, because things like the auto-parenting need to actually find the camera rig objects.
Another issue I saw is with the Simulator: It can't find the needed Sim script that controls it all because that object is in the inaccessible `DontDestroyOnLoad` scene, too.

There are probably more issues when we test all the example scenes. Like I could see our scripts looking up the script to use in case it has been left blank in the inspector become a problem, too.

Basically anything that lives in that inaccessible scene is gone, even when searching from an object that is part of that scene, too.

**Comment by thestonefox** [comment]
@bddckr at the moment the issue we have is when you switch scenes with persist on load checked, if the code is updated to not try and repopulate then you still keep the relevant game object references but the script parenting doesn't happen.

I assume this is because if you don't do the populate game objects thing then the sdk changed event isn't emitted maybe? or OnLoaded isn't called?

So, is there any way we could still do that call on the scene change event?

Effectively using the existing auto childed objects (the script aliases and anything that's using the SDK Object Alias script) and then making sure the loaded set up event is called so those scripts in the new scene also get called?


So in short my idea (if at all possible) is you keep the object references with persist on load from the initial scene but then on switching scene you clean those object references of anything that was attached, know you've switched scene then ensure any other scripts do their setup based on those original references again.

**Comment by bddckr** [comment]
> I assume this is because if you don't do the populate game objects thing then the sdk changed event isn't emitted maybe? or OnLoaded isn't called?

No. The parenting isn't done because the parent is not found as it's in the inaccessible scene.

> So, is there any way we could still do that call on the scene change event?

Sadly not as we can't look up things in that extra scene. We could only get a reference to the needed object via something like `VRTK_SDKManager.instance.scriptAliasLeftController` for example. Because the `scriptAliasLeftController` is already referenced there for us.

This still doesn't solve all the other issues that arise:
- When you switch to the Simulator it tries to find the parent script it needs, which it won't find _because it's in the inaccessible scene_.
- I'm sure there are others like this, plus we are forcing our users to do the same workarounds - if we find one, that is.

---

So again: This issue is not about auto-populating doing something wrong. It's just that the scene is inaccessible, which means we can't get any reference to the objects except when we already have that reference set up before it's moved into the inaccessible scene. (I.e. it's already set in the Editor and saved into the scene with the references set.)

**Comment by bddckr** [comment]
I have a workaround that makes it possible to find objects in that inaccessible scene by using `VRTK_SDKManager.instance.gameObject.scene`, though the issue with that is that the returned scene only works in the Editor. In a standalone build it's an invalid scene and can't be used.

**Comment by bddckr** [comment]
Back from Unite and it looks like we may be after something with [`SceneManager.MoveGameObjectToScene`](https://docs.unity3d.com/ScriptReference/SceneManagement.SceneManager.MoveGameObjectToScene.html). Gonna check this soon.
